### ISSUE TITLE
SPARQL 1.1 Graph Store Protocol operations hang

### ISSUE THREAD

**Issue by jakubklimek** [openlink/virtuoso-opensource#165]
When trying to debug https://github.com/openlink/virtuoso-opensource/issues/128 I probably discovered a separate issue. When issuing multiple CLEAR GRAPH / LOAD GRAPH statements, Virtuoso comes to a state where every CLEAR GRAPH times out eventually. We have the timeout set to 2h - this happens even with graphs that normally take a minute to clear. Meanwhile, it however answers SPARQL queries normally, so it is not a complete crash. <code>/etc/init.d/virtuoso-opensource stop</code> does not work either. In this state, I need to use force-stop.

In virtuoso.log there is
<code><pre>03:00:50 \* Monitor: Locks are held for a long time
03:02:02 \* Monitor: Many lock waits
03:04:03 \* Monitor: Many lock waits
03:06:07 \* Monitor: Many lock waits
03:08:08 \* Monitor: Many lock waits
</code></pre>

**Comment by HughWilliams** [comment]
What is the size of the graphs being cleared/loaded and do you have sample data and steps to recreate ?

What is the memory consumption of Virtuoso and Available free memory as reported by top at this point. Also, what does the "status();" command run from isql report ?

**Comment by jakubklimek** [comment]
Actually... the scenario when this happens regularly is when used with our LOD2 tool UnifiedViews when multiple pipelines start to clear / load graphs using SPARQL 1.1 Graph Store protocol.

<a href="http://internal.opendata.cz/dumps.zip">Our DB (2.4GB)</a>
<a href="http://internal.opendata.cz/ares.zip">Graphs to clear and reload</a>

The size of the manipulated graphs is about 10M triples in total.

Memory consumption: 10GB virtual memory, 7GB physical memory

Log contains the lock waits messages, querying (read-only) works normally, but write operations hang. Virtuoso cannot be stopped politely (by sending signal 15). It logs it, but does not close (or does, but only after hours of waiting)

virtuoso.trx: 270MB, virtuoso-temp.db: 700MB

Status:

<pre><code> OpenLink Virtuoso Server
 Version 07.10.3208-pthreads for Linux as of May 18 2014
 Started on: 2014-05-18 23:20 GMT+2
 
 Database Status:
  File size 0, 1236224 pages, 464896 free.
  680000 buffers, 384423 used, 72740 dirty 2 wired down, repl age 7694900 0 w. io 16 w/crsr.
  Disk Usage: 385425 reads avg 0 msec, 0% r 0% w last 0 s, 262 writes flush 0 MB,
  2257 read ahead, batch = 164. Autocompact 16 in 13 out, 17% saved.
 Gate: 7198 2nd in reads, 0 gate write waits, 0 in while read 0 busy scrap.
 Log = virtuoso.trx, 270168986 bytes
 705015 pages have been changed since last backup (in checkpoint state)
 Current backup timestamp: 0x0000-0x00-0x00
 Last backup date: unknown
 Clients: 369 connects, max 1 concurrent
 RPC: 3187 calls, -369 pending, 1 max until now, 0 queued, 0 burst reads (0%), 0 second 1343M large, 1473M max
 Checkpoint Remap 1842 pages, 0 mapped back. 0 s atomic time.
  DB master 1236224 total 464896 free 1842 remap 281 mapped back
  temp 79104 total 79098 free
 
 Lock Status: 1 deadlocks of which 0 2r1w, 94 waits,
  Currently 26 threads running 20 threads waiting 4 threads in vdb.
 Pending:
  1173521: IEP INTERNAL
  1173522: IEP INTERNAL
  1173523: IEP INTERNAL
  1173524: IEP INTERNAL
  1173525: IEP INTERNAL
  1173526: IEP INTERNAL
  1173527: IEP INTERNAL
  1173528: IEP INTERNAL
  1173529: IEP INTERNAL
  1173530: IEP INTERNAL
  1173531: IEP INTERNAL
  1173532: IEP INTERNAL
  1173533: IEP INTERNAL
  1173534: IEP INTERNAL
  1173535: IEP INTERNAL
  1173536: IEP INTERNAL
  1173537: IEP INTERNAL
  1173538: IEP INTERNAL
  1173539: IEP INTERNAL
  1173540: IEP INTERNAL
  1173541: IEP INTERNAL
  1173542: IEP INTERNAL
  1173543: IEP INTERNAL
  1173544: IEP INTERNAL
  1173545: IEP INTERNAL
  1173546: IEP INTERNAL
  1173547: IEP INTERNAL
  1173548: IEP INTERNAL
  1173549: IER INTERNAL 195.113.17.20
  44: IER INTERNAL
  40: IER INTERNAL
  36: IER INTERNAL
  32: IER INTERNAL
  28: IER INTERNAL
  24: IER INTERNAL
  20: IER INTERNAL
  16: IER INTERNAL
  12: IER INTERNAL
  8: IER INTERNAL
  4: IER INTERNAL
  0: IER INTERNAL
  45: IER 195.113.17.20 Waiting: INTERNAL
  41: IER INTERNAL
  37: IER INTERNAL
  33: IER INTERNAL
  29: IER INTERNAL
  25: IER INTERNAL
  21: IER INTERNAL
  17: IER INTERNAL
  13: IER INTERNAL
  9: IER INTERNAL
  5: IER INTERNAL
  1: IER INTERNAL
  46: IER 195.113.17.20
  42: IER INTERNAL
  38: IER INTERNAL
  34: IER INTERNAL
  30: IER INTERNAL
  26: IER INTERNAL
  22: IER INTERNAL
  18: IER INTERNAL
  14: IER INTERNAL
  10: IER INTERNAL
  6: IER INTERNAL
  2: IER INTERNAL
  43: IER INTERNAL
  39: IER INTERNAL
  35: IER INTERNAL
  31: IER INTERNAL
  27: IER INTERNAL
  23: IER INTERNAL
  19: IER INTERNAL
  15: IER INTERNAL
  11: IER INTERNAL
  7: IER INTERNAL
  3: IER INTERNAL
  1173551: IER NO_CONN NO_CONN 195.113.17.20</code></pre>

**Comment by HughWilliams** [comment]
Downloaded and loaded the dump & ares zip files and loaded them into an empty develop/7 database, deleted and and reloaded the ares.zip graphs twice sucessfully without problems:

SQL> sparql clear graph http://linked.opendata.cz/resource/dataset/ares/basic/metadata;

Done. -- 3 msec.
SQL> sparql clear graph http://linked.opendata.cz/resource/dataset/ares/basic;

Done. -- 257763 msec.
SQL> sparql clear graph http://linked.opendata.cz/resource/dataset/ares/or/metadata;

Done. -- 3 msec.
SQL> sparql clear graph http://linked.opendata.cz/resource/dataset/ares/or;

Done. -- 661319 msec.
SQL> delete from load_list;

Done. -- 0 msec.
SQL> select \* from load_list;
ll_file                                                                           ll_graph                                                                          ll_state    ll_started           ll_done              ll_host     ll_work_time  ll_error
VARCHAR NOT NULL                                                                  VARCHAR                                                                           INTEGER     TIMESTAMP            TIMESTAMP            INTEGER     INTEGER     VARCHAR

---

0 Rows. -- 0 msec.
SQL> ld_dir_all('.', '*.ttl', 'graph_name');

Done. -- 4 msec.
SQL> rdf_loader_run(); &
SQL> Connected to OpenLink Virtuoso
Driver: 07.10.3208 OpenLink Virtuoso ODBC Driver
OpenLink Interactive SQL (Virtuoso), version 0.9849b.
Type HELP; for help and EXIT; to exit.

SQL> rdf_loader_run(); &
SQL> Connected to OpenLink Virtuoso
Driver: 07.10.3208 OpenLink Virtuoso ODBC Driver
OpenLink Interactive SQL (Virtuoso), version 0.9849b.
Type HELP; for help and EXIT; to exit.

SQL> 
Done. -- 41173 msec.

Done. -- 96896 msec.

SQL>

**Comment by jakubklimek** [comment]
Well... executed like this, it is in sequence. I suspect that the problem has to do with parallelism and transactions in a situation where part of virtuoso's memory is in swap. Have you ran it on a machine with 8gb of memory or less? Try clearing and reloading random graphs from the provided dataset in parallel until the lock wait messages start appearing in virtuoso.log and virtuoso starts swapping.

**Comment by jakubklimek** [comment]
May be related to https://github.com/openlink/virtuoso-opensource/issues/312

**Comment by yesonmect** [comment]
@jakubklimek， i met the same problem with you. As i import triple files to virtuoso，the write process is hang，but the read is ok. And in log file，i found the error “many lock waits”，at the same time，i see the same status report using “status（）；”
Can you tell me how to solve it or give me some suggestion？
Thanks very much！

**Comment by HughWilliams** [comment]
@yesonmect: Can you provide your actual `status();` output for review ?

Also, have you performance tuned your Virtuoso server as detailed at https://virtuoso.openlinksw.com/dataspace/doc/dav/wiki/Main/VirtRDFPerformanceTuning
